ocaml/libs/stunnel/stunnel.ml

```diff
@@ -288,54 +299,42 @@ let disconnect ?(wait = true) ?(force = false) x =
 exception Stunnel_initialisation_failed
 
 (* Internal function which may throw Stunnel_initialisation_failed *)
-let with_attempt_one_connect ?unique_id ?(use_fork_exec_helper = true)
-    ?(write_to_log = fun _ -> ()) verify_cert extended_diagnosis host port f =
+let attempt_one_connect ?(use_fork_exec_helper = true)
+    ?(write_to_log = fun _ -> ()) ?(extended_diagnosis = false) data_channel
+    verify_cert host port =
   Unixfd.with_pipe () ~loc:__LOC__ @@ fun config_out config_in ->
   let config_out_uuid = Uuidm.to_string (Uuidm.create `V4) in
   let config_out_fd =
     string_of_int (Unixext.int_of_file_descr Unixfd.(!config_out))
   in
-  let fds_needed =
-    [Unixfd.(!config_out); Unix.stdin; Unix.stdout; Unix.stderr]
-  in
   let configs = [(config_out_uuid, Unixfd.(!config_out))] in
   let args =
     ["-fd"; (if use_fork_exec_helper then config_out_uuid else config_out_fd)]
   in
-  Unixfd.with_socketpair Unix.PF_UNIX Unix.SOCK_STREAM 0 ~loc:__LOC__
-  @@ fun data_in data_out ->
-  (* Dereference just once to ensure we are consistent in t and config_file *)
-  let t =
-    {
-      pid= Nopid
-    ; fd= data_out
-    ; host
-    ; port
-    ; connected_time= Unix.gettimeofday ()
-    ; unique_id
-    ; logfile= ""
-    ; verified= verify_cert
-    }
-  in
-  let result =
+  let start sock_of_stunnel config =
     Forkhelpers.with_logfile_fd "stunnel" ~delete:(not extended_diagnosis)
       (fun logfd ->
         let path = stunnel_path () in
-        let fdops =
-          [
-            Unsafe.Dup2 (Unixfd.(!data_in), Unix.stdin)
-          ; Unsafe.Dup2 (Unixfd.(!data_in), Unix.stdout)
-          ; Unsafe.Dup2 (logfd, Unix.stderr)
-          ]
+        let fds_needed, fdops, sock =
```

Comment:
`sock` is a raw fd here, I think it might be better if we keep it undereferenced and dereference it with `Unixfd.(!)` only at the callsite. It is only used below in Forkhelpers.safe_close_and_exec currently, but in the future we might accidentally attempt to use it elsewhere (and it will get closed by the with_socketpair function, so we might get EBADF in that case).
I'm not sure what we could do at the compiler/typesystem level to ensure the fd is used correctly (OCaml doesn't have linear types built-in), I think for now we relied on this convention of dereferencing the fd wrapper as late as possible and in the smallest scope possible.